File: python/sieve/sieve.py
import math

class Sieve:

    def __init__(self) -> None:
        pass

    def nth_prime(self, n: int) -> int:
        if n < 0:
            raise ValueError("n must be non-negative")

        # Estimate upper bound for nth prime using the prime number theorem
        if n < 6:
            upper_bound = 15
        else:
            # TODO: apparently this is a loose upper bound, other tighter bounds exist
            upper_bound = int(n * (math.log(n) + math.log(math.log(n))))

        # Sieve of Eratosthenes
        # Create a list of index 0 - n of True's
        sieve = [True] * (upper_bound + 1)
        # 0 and 1 are not prime
        sieve[0] = sieve[1] = False
        
        # Start at 2, go to sqrt(n)
        for num in range(2, int(math.sqrt(upper_bound))+1):
            if sieve[num] == True:
                # Mark all multiples False, indexing by num
                for composite_index in range(num * num, upper_bound, num):
                    sieve[composite_index] = False

        # Get list of primes
        prime_numbers = []
        for prime_index in range(0, len(sieve)):
            if sieve[prime_index] == True:
                prime_numbers.append(prime_index)
            
        # Return nth prime
        return prime_numbers[n]



if __name__ == "__main__":
    sieve = Sieve()
    sieve.nth_prime(10)

File: go/pkg/sieve/sieve.go
package sieve

import (
	"math"
)

type Sieve interface {
	NthPrime(n int64) int64
}

type sieveOfEratosthenes struct {
}

func NewSieve() Sieve {
	return &sieveOfEratosthenes{}
}

func (s *sieveOfEratosthenes) NthPrime(n int64) int64 {
	upper_bound := 0
	primes := []int64{}

	// Handle edge case for negative n and return -1 for invalid input
	if n < 0 {
		return -1
	}

	// Estimate the upper limit for the nth prime using the prime number theorem
	if n < 6 {
		upper_bound = 15
	} else {
		float_n := float64(n)
		upper_bound = int(float_n * (math.Log(float_n) + math.Log(math.Log(float_n))))
	}

	// Create boolean array and initialize as true
	sieve_slice := make([]bool, upper_bound+1)
	sieve_slice[0] = false
	sieve_slice[1] = false
	for i := 2; i <= upper_bound; i++ {
		sieve_slice[i] = true
	}

	// Sieve of Eratosthenes
	for num := 2; num <= int(math.Sqrt(float64(upper_bound))); num++ {
		if sieve_slice[num] {
			for composite_index := num * num; composite_index <= upper_bound; composite_index += num {
				sieve_slice[composite_index] = false
			}
		}
	}

	// Get list of primes
	for prime_index := 0; prime_index <= upper_bound; prime_index++ {
		if sieve_slice[prime_index] {
			primes = append(primes, int64(prime_index))
		}
	}

	return primes[n]
}
